### PR TITLE
WIP: Custom error type for secretbox::open and secretbox::open_detached

### PR DESCRIPTION
**Pull request by mso42**
Adds a custom error type `SecretBoxError`. Is this pattern good enough to implement everywhere?

Addresses https://github.com/sodiumoxide/sodiumoxide/issues/221

### CODE REVIEW

File: src/crypto/secretbox/error.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#[cfg(feature = "std")]
+impl std::fmt::Display for SecretBoxError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for SecretBoxError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
```
Comment (mexus) [inline src/crypto/secretbox/error.rs:22]: It's a default implementation, isn't it? https://doc.rust-lang.org/src/std/error.rs.html#100 So the whole block is equivalent to `impl std::error::Error for SecretBoxError {}`

File: src/crypto/secretbox/error.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    /// function, which expects a message prepended with a MAC.
+    CiphertextTooShort,
+    /// The ciphertext failed verification.
+    CiphertextFailedVerification,
+}
+
+#[cfg(feature = "std")]
+impl std::fmt::Display for SecretBoxError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
```
Comment (mexus) [inline src/crypto/secretbox/error.rs:16]: Normally, the `Display` trait is used to provide human-readable error messages, target at end users. Won't it make better sense to display what is currently written in doc comments instead?

File: src/crypto/secretbox/error.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#[derive(Debug, PartialEq)]
+pub enum SecretBoxError {
+    /// The ciphertext was too short. This results from using the `crypto_secretbox_open_easy`
+    /// function, which expects a message prepended with a MAC.
+    CiphertextTooShort,
+    /// The ciphertext failed verification.
+    CiphertextFailedVerification,
+}
+
+#[cfg(feature = "std")]
```
Comment (mexus) [inline src/crypto/secretbox/error.rs:13]: Since `std::fmt::Display` is an alias for `core::fmt::Display`, it is possible to provide better user experience for no-std users by implementing the latter trait unconditionally, even though the `Error` trait could not be implemented with absence of std

**Comment by mso42** [comment]
All good feedback, thanks

File: src/crypto/secretbox/error.rs (hunk 2 above)
Comment (mso42) [reply src/crypto/secretbox/error.rs:16]: Agreed. Do you know of an easy way to do this wthat doesn't involve duplicating the text in the doc comments?

Comment (mexus) [reply src/crypto/secretbox/error.rs:16]: The only way I'm aware of is to use a custom macro :) like that one: [displaydoc](https://crates.io/crates/displaydoc) crate

**Comment by mso42** [comment]
New commit resolving the above comments and using `displaydoc` crate to derive Display for the error.

Adding a dependency is somewhat controversial I guess, but:
- it has minimal child dependencies (standard macro stuff like `proc_macro2`, `syn`, `quote`)
- it works in `no-std`

The alternative is to duplicate the display strings in the documentation comments and the `Display` implementation.

**Comment by kpp** [comment]
> New commit resolving the above comments and using displaydoc crate to derive Display for the error.

Looks like we are not far away from snafu :)

**Comment by dnaq** [comment]
Yeah, I feel like a whole new dependency for not having to write a display
implementation is a bit overkill.

**Comment by mexus** [comment]
> > New commit resolving the above comments and using displaydoc crate to derive Display for the error.
> 
> Looks like we are not far away from snafu :)

snafu supports no_std btw :) and with its default features turned off it introduces only syn/quote/proc-macro2 and [doc-comment macro](https://docs.rs/doc-comment) crates as additional dependencies.

> Yeah, I feel like a whole new dependency for not having to write a display implementation is a bit overkill.

It definitely looks like an overkill when writing a display implementation only once or twice, but what will it look like when more error handling to be introduced?

**Comment by mso42** [comment]
I think that including the entire docstring as the Display message for the error may not always be the right thing to do. For example, if the docstring (for some reason) were multiple lines with code snippets etc. etc.

My preference is to keep error messages reasonably short, so that they can be pasted into search engines and relevant documentation found from there. This is why I simply used the Debug string for the Display in the initial patch.

I note that in our case, our error types are probably not going to be that complicated. (Probably just enums with no fields.) And in that case, I think something like snafu or thiserror is even more overkill than displaydoc.

**Comment by mso42** [comment]
I have got rid of the display-doc stuff, and written some error messages that say nothing more than the enum field, but in plain English.

Also squashed all the commits so far.

**Comment by mso42** [comment]
While we are still pending agreement on the actual text of the error messages, can I get feedback on the general approach here, i.e. having one error type per type of operation?

Or would it be better to have one big `SodiumOxideError` type at the crate root and have everything which could error return instances of that.

My preference is for the former, so that handling code doesn't have to handle `pwhash`-related errors if it is calling `sealedbox` functions (for example).

The argument in favour of the latter is that it keeps things simpler within the crate I guess.

**Comment by dnaq** [comment]
One error type at the crate root. Much better usability.

**Comment by mso42** [comment]
So calling code would have to do:

```rust
match secretbox::seal_detached(&mut buf, &nonce, &key) {
    Ok(tag) => { /* ... */ },
    Err(SodiumError::SecretBoxSealError) => { /* handle error */ },
    Err(_) => { /* how could this possibly even happen? what do we do here? panic? */ },
}
```

I don't see how that is better usability.

**Comment by dnaq** [comment]
For any library/binary that doesn’t want to handle the error it becomes
something like:

```
enum Error {
  Foo(...),
  Bar(...),
  CryptoError(sodiumoxide::Error),
}
```

**Comment by mso42** [comment]
Oh yes that is true. What do other libraries do?

**Comment by dnaq** [comment]
Most that I’ve seen, reqwest is one.

**Comment by adamreichold** [comment]
> So calling code would have to do:

Another thing to keep in mind is that while SNAFU-style nested error enums can allow exhaustive matching, this can imply problems with backwards compatibility. Meaning we might want to make all our error enums `#[non_exhaustive]` so that even the specialised enums could not be exhaustively and would always need a "unexpected fatal error" match arm.

**Comment by kpp** [comment]
@mso42 why did you close this PR?..

**Comment by mso42** [comment]
Oh, I didn't mean to. I renamed my working branch and reset my repository's `master` to match upstream. Looks like that auto-closed the PR.

I have done a little work on this, and intend to do more, but it is a big job and I'm not sure I have the motivation to see it through. If I get to something that I can believably call not a work-in-progress, I'll reopen another PR.
